### PR TITLE
Mixed detector

### PR DESCRIPTION
**Pull request by rhiannonlynne**
Fixed m2throughput.dat file and ipython notebook to generate this file.

Added SEDs. 

Added sedUtils and bandpassUtils - note bandpassUtils will let you build throughput curves from base components! (ie.. they duplicate some of the functionality of the ipython notebooks and we should probably think about what to do with that). 

Added mixedDetector analysis (uses bandpassUtils and sedUtils, requires sims_photUtils).

### CODE REVIEW

**Comment by rhiannonlynne** [comment]
I'd like to have @cclaver review this, and also Scott Daniels (@danielsf)

File: python/bandpassUtils.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+#       mirror[1,2,3] / m[1,2,3]Losses.dat, m[1,2,3]ProtAlIdeal.dat   (multiply ideal and loss for each mirror, multiply all mirrors)
+
+def setDefaultDirs(rootDir=None):
+    """
+    Returns a dictionary with the default directory locations of each component of the system throughput.
+    """
+    # Set SYSENG_THROUGHPUTS_DIR to the root dir for syseng_throughputs ('setup syseng_throughputs' will do this automatically)
+    defaultDirs = {}
+    if rootDir is None:
+        rootDir = os.getenv('SYSENG_THROUGHPUTS_DIR')
```
Comment (danielsf) [inline python/bandpassUtils.py]: Do you want to support people who have not use eups to setup the stack?  If not, should this be

rootDir = lsst.utils.getPackageDir('sysengh_throughputs')

?

File: python/bandpassUtils.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+    defaultDirs['filters'] = os.path.join(rootDir, 'components/camera/filters')
+    for mirror in ('mirror1', 'mirror2', 'mirror3'):
+        defaultDirs[mirror] = os.path.join(rootDir, 'components/telescope', mirror)
+    defaultDirs['atmosphere'] = os.path.join(rootDir, 'siteProperties')
+    return defaultDirs
+
+
+def buildVendorDetector(vendorDir, addLosses=True):
+    """
+    Assumes there is a file *QE.dat and *Losses.dat to represent the QE response and the Losses.
```
Comment (danielsf) [inline python/bandpassUtils.py]: This docstring should be more explicit, i.e. "if addLosses==True, multiplies losses and QE together in restuls; if addLosses==False, only returns QE"

File: python/bandpassUtils.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        lossfiles = glob(os.path.join(vendorDir, '*Losses.dat'))
+        loss = Bandpass()
+        loss.readThroughputList(lossfiles)
+        wavelength, sb = qe.multiplyThroughputs(loss.wavelen, loss.sb)
+        qe.setBandpass(wavelength, sb)
+    return qe
+
+def buildGenericDetector(detectorDir, addLosses=True):
+    """
+    Combine detector throughputs from multiple vendors, using the minimum value of QE*losses at each wavelength.
```
Comment (danielsf) [inline python/bandpassUtils.py]: mention behavior of addLosses in docstring

File: python/bandpassUtils.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    sbAll = []
+    for vendorDir in vendorDirs:
+        qe = buildVendorDetector(vendorDir, addLosses=addLosses)
+        sbAll.append(qe.sb)
+    wavelen = qe.wavelen
+    sbMin = (np.array(sbAll)).min(axis=0)
+    qe.setBandpass(wavelen, sbMin)
+    return qe
+
+def buildFilters(filterDir, addLosses=True):
```
Comment (danielsf) [inline python/bandpassUtils.py:152]: mention behavior of addLosses in docstring

File: python/bandpassUtils.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        # Read and multiply the losses.
+        lossfiles = glob(os.path.join(filterDir, 'filterLosses', '*.dat'))
+        loss = Bandpass()
+        loss.readThroughputList(lossfiles)
+        for f in filters:
+            wavelen, sb = filters[f].multiplyThroughputs(loss.wavelen, loss.sb)
+            filters[f].setBandpass(wavelen, sb)
+    return filters
+
+def savitzky_golay(y, window_size=31, order=3, deriv=0, rate=1):
```
Comment (danielsf) [inline python/bandpassUtils.py:174]: Should this method have a docstring?

Comment (danielsf) [inline python/bandpassUtils.py:174]: I it is not clear to me what deriv and rate are

File: python/bandpassUtils.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+    b = np.asmatrix([[k**i for i in order_range] for k in range(-half_window, half_window+1)])
+    m = np.linalg.pinv(b).A[deriv] * rate**deriv * np.math.factorial(deriv)
+    # pad the signal at the extremes with
+    # values taken from the signal itself
+    firstvals = y[0] - np.abs(y[1:half_window+1][::-1] - y[0] )
+    lastvals = y[-1] + np.abs(y[-half_window-1:-1][::-1] - y[-1])
+    y = np.concatenate((firstvals, y, lastvals))
+    return np.convolve(m[::-1], y, mode='valid')
+
+def buildLens(lensDir, addLosses=True):
```
Comment (danielsf) [inline python/bandpassUtils.py:202]: mention behavior of addLosses in docstring

File: python/bandpassUtils.py (hunk 5 above)
Comment (rhiannonlynne) [inline python/bandpassUtils.py:174]: This whole function was lifted from Chuck's ipython notebook. I actually don't know what it's supposed to do, other than smooth the wavelength response of the borosilicate glass throughput curve (and I don't know why you have to do that). I have a request to Chuck to clarify and to explain, but otherwise this whole function is just here because it looked like I needed it (see makeLens*.ipynb). Perhaps @cclaver can fill in some more information on this PR?

File: python/bandpassUtils.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+    # Read the broad band antireflective (BBAR) coatings files.
+    bbarfiles = glob(os.path.join(lensDir, 'l*BBAR.dat'))
+    bbars = Bandpass()
+    bbars.readThroughputList(bbarfiles)
+    # Multiply the bbars by the glass.
+    wavelen, sb = lens.multiplyThroughputs(bbars.wavelen, bbars.sb)
+    lens.setBandpass(wavelen, sb)
+    # Add losses.
+    if addLosses:
+        lossfiles = glob(os.path.join(lensDir, 'l*Losses", ''*.dat'))
```
Comment (danielsf) [inline python/bandpassUtils.py]: Does this line work?  I think the double- and single- quotation marks got mixed up.

File: python/bandpassUtils.py (hunk 6 above)
Comment (danielsf) [inline python/bandpassUtils.py:202]: Does the loss file sub-directory need to obey a specific naming convention?

File: python/bandpassUtils.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+        loss = Bandpass()
+        loss.readThroughputList(lossfiles)
+        wavelen, sb = lens.multiplyThroughputs(loss.wavelen, loss.sb)
+        lens.setBandpass(wavelen, sb)
+    return lens
+
+def buildMirror(mirrorDir, addLosses=True):
```
Comment (danielsf) [inline python/bandpassUtils.py:236]: mention addLosses in docstring

File: python/bandpassUtils.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+    """
+    Read the atmosphere throughput curve.
+    Returns a bandpass object.
+    """
+    atmofile = os.path.join(atmosDir, 'pachonModtranAtm_12.dat')
+    atmo = Bandpass()
+    atmo.readThroughput(atmofile)
+    return atmo
+
+def buildHardwareAndSystem(defaultDirs, addLosses=True):
```
Comment (danielsf) [inline python/bandpassUtils.py:268]: docstring should specify what defaultDirs is (a dict pointing to the directories holding the data files associated with each throughput component)

File: python/bandpassUtils.py [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+    for f in filters:
+        hardware[f] = Bandpass()
+        system[f] = Bandpass()
+        wavelen = filters[f].wavelen
+        hw_sb = core_sb * filters[f].sb
+        hardware[f].setBandpass(wavelen, hw_sb)
+        system[f].setBandpass(wavelen, hw_sb*atmosphere.sb)
+    return hardware, system
+
+def plotBandpasses(bandpassDict, title=None, newfig=True, savefig=False, addlegend=True,
```
Comment (danielsf) [inline python/bandpassUtils.py:310]: Define arguments and kwargs in docstring

File: python/bandpassUtils.py (hunk 7 above)
Comment (rhiannonlynne) [inline python/bandpassUtils.py]: It did somehow (results of the bandpass look like the version calculated by Chuck, so must include losses), but I will fix.

File: python/bandpassUtils.py (hunk 2 above)
Comment (rhiannonlynne) [inline python/bandpassUtils.py]: Yeah - needs more documentation in general. Will add.

File: python/bandpassUtils.py (hunk 1 above)
Comment (rhiannonlynne) [inline python/bandpassUtils.py]: Yes, I will swap to this.

File: python/sedUtils.py [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+
+
+def readPhotSeds(sedDir=None):
+    """Read all the seds provided by this package, storing them in a
+    dictionary and saving lists of the SEDs of each type (so that
+    they can be separated later if desired). """
+    # The environment variable PHOT_REF_SEDS_DIR is set if the file
+    #  'phot_ref.csh' is sourced.
+    if sedDir == None:
+        sedDir = os.path.join(os.getenv('SYSENG_THROUGHPUTS_DIR'), 'seds')
```
Comment (danielsf) [inline python/sedUtils.py]: Here's another instance where we should probably use lsst.utils.getPackageDir

File: python/sedUtils.py [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+                    sedDicts[newsedname] = redshiftSingleSED(sedDict[s], z)
+                    newSedsForList.append(newsedname)
+                sedlists[objtype] += newSedsForList
+    return sedDict, sedlists
+
+
+def redshiftSingleSED(sed_in, z):
+    # Make a copy of the input SED, as otherwise we will overwrite original arrays.
+    sed_out = deepcopy(sed_in)
+    sed_out.redshiftSED(z, dimming=False)
```
Comment (danielsf) [inline python/sedUtils.py:83]: what is the rationale behind setting dimming=False?  In the catalogs, we setting dimming=True

File: python/sedUtils.py [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+    # Read the files.
+    sedDict = {}
+    for filename, s in zip(inputfileList, sedlists['any']):
+        sedDict[s] = Sed()
+        sedDict[s].readSED_flambda(filename)
+    # Return the sed dictionary and the dictionary containing the lists of each type of object.
+    return sedDict, sedlists
+
+
+def makeRedshiftedSeds(sedDict, sedlists, redshifts):
```
Comment (danielsf) [inline python/sedUtils.py]: it looks like this method returns two dicts: one (sedDict) that is keyed on the SEDs individual name.  One (sedlists) that is keyed on the object type and whose values are a list of all the SEDs of that object type.  sedlists should be renamed so it is obvious that it is a dict.

File: python/sedUtils.py [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+
+
+def redshiftSingleSED(sed_in, z):
+    # Make a copy of the input SED, as otherwise we will overwrite original arrays.
+    sed_out = deepcopy(sed_in)
+    sed_out.redshiftSED(z, dimming=False)
+    return sed_out
+
+
+def matchSedsBp(sedDict, bpDict, refFilter=None):
```
Comment (danielsf) [inline python/sedUtils.py:87]: Should this method make use of Sed.setupPhiArray to further speed up the calculation?

Comment (danielsf) [inline python/sedUtils.py:87]: There is now a class in sims_photUtils called BandpassDict which reads in a bunch of bandpasses, and then automatically resamples them to the same wavelength grid and calculates their phi array.  It uses Sed.manyMagCalc to calculate the magnitude of any Sed in all of the bandpasses simultaneously.  Maybe that is worth using here?

File: python/sedUtils.py [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+            bpDict[f].resampleBandpass(wavelen_min=wavelen_match.min(), wavelen_max=wavelen_match.max(),
+                                       wavelen_step = wavelen_match[1] - wavelen_match[0])
+    # Check all seds in sedDict match in wavelength space.
+    for s in sedDict.keys():
+        if sedDict[s].needResample(wavelen_match=wavelen_match):
+            sedDict[s].resampleSED(wavelen_match=wavelen_match)
+    return sedDict, bpDict
+
+
+def calcNatMags(bandpassDict, sedDict, sedlists):
```
Comment (danielsf) [inline python/sedUtils.py]: This method does not appear to use sedDict

Comment (danielsf) [inline python/sedUtils.py]: wait.  Nevermind.  I see it.  I am very confused what the difference between sedDict and sedLists is.

File: python/sedUtils.py (hunk 13 above)
Comment (danielsf) [inline python/sedUtils.py]: Okay: sedlists is a dict of lists of sed names.  Is it possible to just work with discrete sedDicts (one for quasars, one for galaxies, one for stars).  Or even just make sedDict a dict of dicts, so that

sedDict ['quasar']['quasar_name'] 

was a quasar SED?  That way, you wouldn't have to keep carrying around both sedDict and sedlists

File: python/mixedDetector_photometry.py [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+import matplotlib.pyplot as plt
+from lsst.sims.photUtils import Bandpass, Sed, PhotometricParameters, LSSTdefaults, SignalToNoise
+import bandpassUtils as bu
+import sedUtils as su
+
+def calcM5s(hardware, system, title='m5'):
+    photParams = PhotometricParameters()
+    lsstDefaults = LSSTdefaults()
+    darksky = Sed()
+    darksky.readSED_flambda(os.path.join('../siteProperties', 'darksky.dat'))
```
Comment (danielsf) [inline python/mixedDetector_photometry.py:17]: This should also use lsst.utils.getPackageDir so that it does not depend on where it is being run from

**Comment by danielsf** [comment]
I am effectively done with my review.

File: python/sedUtils.py (hunk 13 above)
Comment (rhiannonlynne) [inline python/sedUtils.py]: I think this makes sense. I had some other reason for doing this originally (a long time ago, when I was using these methods for the first time, but I maybe wasn't so smart at python then and anyway I don't remember why I did it in the first place .. but now is a good time to clean this up).

**Comment by rhiannonlynne** [comment]
@cclaver any comments? I'd  like to get this merged to master.

**Comment by rhiannonlynne** [comment]
Talked to @cclaver via email. Incorporated changes suggested by @danielsf and added some new python scripts which would be useful. 
@cclaver - note you can run "python testComponents.py" to check updates of the components vs. old versions - the old "intermediate files" are still present, but separated into an "IntermediateFiles" directory.
